central/group/datastore/datastore_impl.go

```diff
@@ -105,29 +170,94 @@ func (ds *dataStoreImpl) Mutate(ctx context.Context, remove, update, add []*stor
 		return sac.ErrResourceAccessDenied
 	}
 
-	for _, grp := range append(remove, update...) {
-		if err := ValidateGroup(grp); err != nil {
+	// Lock to ensure that all mutations happen as one
+	ds.lock.Lock()
+	defer ds.lock.Unlock()
+
+	for _, group := range add {
+		if err := ValidateGroup(group); err != nil {
 			return errox.InvalidArgs.CausedBy(err)
 		}
+
+		if err := setGroupIdForNewGroup(group); err != nil {
+			return err
+		}
+
+		defaultGroup, err := ds.getDefaultGroupForProps(ctx, group.GetProps())
+		if err != nil {
+			return err
+		}
+
+		// Check whether the to-be-added group is a default group, ensure that it does not yet exist.
+		if defaultGroup != nil {
+			return errox.AlreadyExists.Newf("a default group already exists for auth provider %q",
+				group.GetProps().GetAuthProviderId())
+		}
+	}
+	if len(add) > 0 {
+		if err := ds.storage.UpsertMany(ctx, add); err != nil {
+			return err
+		}
 	}
 
-	for _, grp := range add {
-		if err := ValidateGroup(grp); err != nil {
+	for _, group := range update {
+		if err := ValidateGroup(group); err != nil {
 			return errox.InvalidArgs.CausedBy(err)
 		}
 
-		if grp.GetProps().GetId() != "" {
-			return errox.InvalidArgs.Newf("id should be empty but %q was provided", grp.GetProps().GetId())
+		// TODO(ROX-11592): Once the deprecation of retrieving groups by their properties is fully deprecated, this condition
+		// can be removed and groups shall only be retrievable via their id.
+		if group.GetProps().GetId() == "" {
+			id, err := ds.findPropsIdByProps(ctx, group.GetProps())
+			if err != nil {
+				return err
+			}
+			// Use the id of the retrieved group to update
+			group.GetProps().Id = id
+		}
+
+		defaultGroup, err := ds.getDefaultGroupForProps(ctx, group.GetProps())
+		if err != nil {
+			return err
+		}
+
+		// Only disallow update of a default group if it does not update the existing default group, if there is any.
+		if defaultGroup != nil && defaultGroup.GetProps().GetId() != group.GetProps().Id {
+			return errox.AlreadyExists.Newf("a default group already exists for auth provider %q",
+				group.GetProps().GetAuthProviderId())
 		}
```

Comment:
I refactored the whole validation and prep part of add and remove so that it can be used in Add, Remove and Mutate. 

I left Remove as-is because the only benefit would be with the fetch by props block which will go away in 73.